rs/replicated_state/src/canister_state/queues.rs

```diff
@@ -174,78 +124,69 @@ pub struct NewCanisterQueues {
 pub struct CanisterOutputQueuesIterator<'a> {
     /// Priority queue of non-empty output queues. The next message to be popped
     /// / peeked is the one at the head of the first queue.
-    queues: VecDeque<(&'a CanisterId, &'a mut OutputQueue)>,
-
-    /// Number of messages that can be popped before the iterator finishes.
-    size: usize,
+    queues: VecDeque<(&'a CanisterId, &'a mut CanisterQueue)>,
 
-    /// The canister's memory usage stats, to be updated as messages are popped.
-    memory_stats: &'a mut MemoryUsageStats,
+    pool: &'a mut MessagePool,
 
-    /// Canister output queue stats, to be updated as messages are popped.
-    queue_stats: &'a mut OutputQueuesStats,
+    /// Number of (potentially stale) messages left in the iterator.
+    size: usize,
 }
 
 impl<'a> CanisterOutputQueuesIterator<'a> {
     fn new(
-        queues: &'a mut BTreeMap<CanisterId, (InputQueue, OutputQueue)>,
-        memory_stats: &'a mut MemoryUsageStats,
-        queue_stats: &'a mut OutputQueuesStats,
+        queues: &'a mut BTreeMap<CanisterId, (CanisterQueue, CanisterQueue)>,
+        pool: &'a mut MessagePool,
     ) -> Self {
         let queues: VecDeque<_> = queues
             .iter_mut()
-            .filter(|(_, (_, queue))| queue.num_messages() > 0)
+            .filter(|(_, (_, queue))| queue.len() > 0)
             .map(|(canister, (_, queue))| (canister, queue))
             .collect();
         let size = Self::compute_size(&queues);
 
-        CanisterOutputQueuesIterator {
-            queues,
-            size,
-            memory_stats,
-            queue_stats,
-        }
+        CanisterOutputQueuesIterator { queues, pool, size }
     }
 
-    /// Returns a reference to the message that `pop` / `next` would return.
+    /// Returns the first message from the next queue.
     pub fn peek(&self) -> Option<&RequestOrResponse> {
-        if let Some((_, queue)) = self.queues.front() {
-            let msg = queue.peek().expect("Empty queue in iterator");
-            return Some(msg);
-        }
-        None
+        let item = self.queues.front()?.1.peek().unwrap();
+
+        let msg = self.pool.get(item.id());
+        assert!(msg.is_some(), "stale reference at the head of output queue");
```

Comment:
The defensive way to program this would be to just pop until we find a non-stale entry, and forgo the invariant. Then the assert can be a debug assert or dropped altogether.

Reply:
It is about a canister having one or more output queues filled to capacity with stale references; and no non-stale message in its output queues. This will prevent any new messages from being enqueued; but will also cause the canister to never be picked up by an `OutputIterator` (because it has no actual output messages), so the stale references are never consumed.

Added to the doc comment of `canister_queues`, where the invariant is defined.

Reply:
It's [here](https://github.com/dfinity/ic/pull/879/files#diff-70106b2012fdd6d4070e7f99bca982eed116cc9e36a93c862d07efaa97793f8dR83).

And no worries about trying to remember everything. We'll all be in the same situation in 6 months' time, so it's better if we document these things now.

Thanks for pointing them out.